In [21]:
import random

def run_simulation(grid_size, max_steps):
    x, y = 0, 0
    goal = (grid_size - 1, grid_size - 1)
    
    for step in range(1, max_steps + 1):
        move = random.choice(['up', 'down', 'left', 'right'])
        
        # Movement logic with boundary checks
        if move == 'up' and y < grid_size - 1:
            y += 1
        elif move == 'down' and y > 0:
            y -= 1
        elif move == 'right' and x < grid_size - 1:
            x += 1
        elif move == 'left' and x > 0:
            x -= 1
            
        if (x, y) == goal:
            return f"SUCCESS: Goal reached in {step} steps."
            
    return f"FAILURE: Goal not reached within {max_steps} steps."

# --- Experiment 1: 10x10 Grid ---
print("--- Grid Size 10 (Goal: 9,9) ---")
for s in [20, 30, 40, 50, 100]:
    print(f"Limit {s}: {run_simulation(10, s)}")

# --- Experiment 2: 21x21 Grid ---
print("\n--- Grid Size 21 (Goal: 20,20) ---")
for s in [100, 200, 500, 1000, 2000]:
    print(f"Limit {s}: {run_simulation(21, s)}")

--- Grid Size 10 (Goal: 9,9) ---
Limit 20: FAILURE: Goal not reached within 20 steps.
Limit 30: FAILURE: Goal not reached within 30 steps.
Limit 40: FAILURE: Goal not reached within 40 steps.
Limit 50: FAILURE: Goal not reached within 50 steps.
Limit 100: FAILURE: Goal not reached within 100 steps.

--- Grid Size 21 (Goal: 20,20) ---
Limit 100: FAILURE: Goal not reached within 100 steps.
Limit 200: FAILURE: Goal not reached within 200 steps.
Limit 500: FAILURE: Goal not reached within 500 steps.
Limit 1000: SUCCESS: Goal reached in 702 steps.
Limit 2000: SUCCESS: Goal reached in 1114 steps.


Assignment: Random Agent Grid Navigation

1. Project Objective
The goal is to simulate a random agent in two different grid environments (10x10 and 21x21) and observe how the maximum step limit affects the probability of the agent reaching the goal.

2. Methodology
Agent: A random mover that selects one of four directions (Up, Down, Left, Right) at each step.

Start Point: (0, 0)

Testing: We increased the step limits incrementally for both a small grid and a large grid to find the "success threshold."

3. Results & Observations

Case 1: 10x10 Grid (Goal at 9,9)

20–100 Steps: In my specific simulation run, the agent failed to reach the goal even with 100 steps. This shows that even in a small grid, a random agent can get "stuck" pacing back and forth near the starting area.

Case 2: 21x21 Grid (Goal at 20,20)

100–500 Steps: The agent failed consistently.

1000 Steps: SUCCESS. The agent reached the goal in 702 steps.

2000 Steps: SUCCESS. The agent reached the goal in 1114 steps.

4. Conclusion
My results prove that success for a random agent is entirely based on luck. Interestingly, my agent found the goal in the larger grid (702 steps) but couldn't find it in the smaller grid within 100 steps. This confirms that without an intelligent algorithm, movement is inefficient and unpredictable.